### ISSUE TITLE
question: Do you plan to add adaptive mesh refine in future?

### ISSUE THREAD

**Issue by ZhouleiJoeStone** [IBirdSoft/openfpm_pdata_2.0.0#3]
Hello: 
I want to know will OpenFPM be able to support adaptive mesh refine like [AMRex](https://amrex-codes.github.io/) in future?

**Comment by incardon** [comment]
Yes. The CPU part is "kind" of ready. The GPU part is on going. Now the umbrella of AMR is pretty big. AMRex support only patch based refinement. In the big picture,  we are planning to support tree and tree-less based refinement and patch refinement, both  CPU and GPU. From the particles side is planned particles with adaptive radius on CPU and GPU. Because is in the original design of OpenFPM.  Time wise ... it depend on the releasing process and work power we have. A preview on some part if there is interest is possible, but production quality test with good documentation and a real case ... it will require more time. From the GPU, we are working on the structures itself at the moment.

**Comment by ZhouleiJoeStone** [comment]
thank you.
Which folder/files should I read first if I want to preview?

**Comment by incardon** [comment]
Is not in github repo, but is on our development repo. But because is public can be browsed here

https://git.mpi-cbg.de/openfpm/openfpm_pdata/tree/amr_struct/src/Amr

The unit tests BOOST_TEST_CASE should show how it work.(Consider it as preview)

Select the branch amr_struct

**Comment by ZhouleiJoeStone** [comment]
thank you @incardon
